whylog/config/rule.py

```diff
@@ -57,6 +65,279 @@ def get_causes_parsers(self):
+            return self.NO_RANGE
+        parser_ranges = self._calculate_parsers_ranges(effect_clues, group_number, group_type)
+        if self._linkage == self.LINKAGE_OR:
+            self._update_parser_ranges_with_or_linkage(
+                effect_clues, group_number, group_type, parser_ranges
+            )
+        return self._aggregate_by_log_type(parser_ranges)
+
+    def _calculate_parsers_ranges(self, effect_clues, group, group_type):
+        """
+        This method calculate search range for every cause parser in rule.
+        This algorithm working only on delta constraints these allow reasoning about
+        parser's search range.
+        Delta constraints analysis:
+            Such constraint contains clues_groups list that contains two elements. Each of
+            elements is a two element list.
+            For example:
+                clues_groups = [[1,1], [0,1]]
+                It means that first group of first parser depend to first group of zero parser.
+                If these first groups of these parsers are in primary keys, we can predict
+                first parser search ranges basing on zero parser search range.
+                Assume that:
+                params = { min_delta: 10, max_delta: 100}
+                zero parser search ranges = {'int': {LEFT_BOUND: 200, RIGHT_BOUND: 250}
+                then
+                first parser search range = {'int': {LEFT_BOUND: 200 - 100, RIGHT_BOUND: 250 - 10}
+                In code first parser is called depended parser, zero parser is called base parser.
+        Connectivity parsers graph is directed graph where two parsers from rule are connected when
+        are connected by delta constraint. This edge is directed from base to depended parser.
+        This implementation assume that connectivity parsers graph is a tree.
+        Algorithm steps:
+            1. Calculate effect parser search range. Take primary key values from effect_clues.
+            2. For every parser in rule find its all delta constraints where is a base parser
+                (_aggregate_constraints method). It is connectivity parsers graph creator.
+            3. Do a BFS order over connectivity parsers graph. Start from effect parser which is a root in tree.
+                Calculate parser search range for every child parser of actual parser if it is possible.
+            4. Set maximal search range for all unreachable parser from effect parser in graph
+            5. Pop effect parser search range.
+        """
+        parser_ranges = {
+            self.EFFECT_NUMBER: self._get_effect_range(effect_clues, group, group_type)
+        }
+        queue = deque([self.EFFECT_NUMBER])
+        aggregated_constraints = self._group_constraints_by_base_parsers()
+        used_parsers = set([self.EFFECT_NUMBER])
+        while queue:
+            parser_number = queue.popleft()
+            for constraint in aggregated_constraints[parser_number]:
```

Comment:
Please put contents of this loop in a separate function to make it easier to comprehend